src/docs/binding-generics.md

```diff
@@ -437,3 +437,191 @@ We could also consider solving this on the runtime level.
+
+A Swift protocol can be mapped to C\# as an interface. Then Swift generic parameters with protocol constraint can be mapped into C\# as generic parameters with interface constraint. This should work for both functions and types.
+
+### Protocol constraints on functions
+
+Consider the following simple example in Swift:
+
+```swift
+public protocol Printable {
+ func printMe()
+}
+
+public func printPrintable<T: Printable>(data: T) {
+ data.printMe()
+}
+
+// Equivalent to: public func printPrintable<T>(data: T) where T : Printable
+```
+
+When a function is declared with a generic argument, an extra implicit argument is added to the end representing the type metadata of the generic argument (see above). If that generic argument is constrained by one or more protocols, the type metadata is followed by a pointer to the protocol witness table for each protocol. In the case of multiple protocols, e.g. T: P3 & P1 & P2, the protocol witness tables are ordered lexically by fully qualified name of the protocol. For detailed description of Metadata accessor refer to [runtime metadata doc](runtime-metadata.md).
+
+This can be projected into C\# as follows:
+
+```csharp
+public interface IPrintableProtocol
+{
+    public void PrintMe();
+    public abstract static NativeHandle GetIPrintableProtocolWitnessTable();
+
+}
+
+public static void PrintPrintable<T>(T data) where T : ISwiftObject, IPrintableProtocol
+{
+    var metadata = TypeMetadata.GetTypeMetadataOrThrow<T>();
+    var protocolWitnessTable = T.GetIPrintableProtocolWitnessTable();
+    PInvokePrintPrintable(data.GetPayload(), metadata, protocolWitnessTable);
+}
+
+[DllImport(Path, EntryPoint = "...")]
+[UnmanagedCallConv(CallConvs = [typeof(CallConvSwift)])]
+private static extern void PInvokePrintPrintable(NativeHandle data, TypeMetadata metadata, NativeHandle protocolWitnessTable);
+```
+
+Then a type implementing the interface could look like:
+
+```csharp
+[StructLayout(LayoutKind.Sequential, Size = /* Extract size */)]
+struct FrozenStruct : ISwiftObject, IPrintableProtocol
+{
+    // Other members
+
+    public static NativeHandle GetIPrintableProtocolWitnessTable()
+    {
+        // This method extracts the witness table symbol from the dynamic library and should be cached.
+        return GetWitnessValueTable("...");
+    }
+
+    public void PrintMe()
+    {
+        var self = new SwiftSelf<FrozenStruct>(this);
+        PInvokePrintMe(self);
+    }
+}
+```
+
+### Protocol constraints on types
+
+Adding constraints on a type's generic parameters is also straightforward, with small adjustments necessary to the initializer and metadata accessor functions. Let's revisit the `Pair` example:
+
+```swift
+@frozen 
+public struct PairNonFrozen<T: Printable, U: Printable> { 
```

Comment:
For the sake of readability - could you add a comment why this is non-frozen. I know we've discussed that `@frozen` doesn't actually mean value-type. So it's possible a better name to use `PairRefType` or something like that, not sure if that's a thing in Swift world as a term though. In any case, I don't understand what makes this a non-frozen thing, is the fact that it has a field of a generic type (so that could be a value type or ref type)?
Would that also be the case if it didn't have fields of those types (and just used the generic parameter in code).

Reply:
Sorry - this was a bad naming leaking from my experimentation code. Now it should be correct.